src/test/java/com/google/cloud/spanner/pgadapter/python/PythonTransactionTests.java

```diff
@@ -0,0 +1,1095 @@
+            .addValues(Value.newBuilder().setStringValue(name).build())
+            .build());
+    return resultSetBuilder.build();
+  }
+
+  @Test
+  public void testSimpleStatementsUsingTransactions() throws IOException, InterruptedException {
+    List<String> statements = new ArrayList<>();
+    String sql1 = "Select * from some_table";
+    String sql2 = "insert into some_table(col1, col2) values(value1, value2)";
+    String sql3 = "Select * from some_table where some_col = some_val";
+
+    statements.add("query");
+    statements.add(sql1);
+
+    statements.add("update");
+    statements.add(sql2);
+
+    statements.add("query");
+    statements.add(sql3);
+
+    statements.add("transaction");
+    statements.add("commit");
+
+    String sql4 = "Select * from some_table3";
+    String sql5 = "insert into some_table3(col1, col2) values(value1, value2)";
+    String sql6 = "Select * from some_table3 where some_col = some_val";
+
+    statements.add("query");
+    statements.add(sql4);
+
+    statements.add("update");
+    statements.add(sql5);
+
+    statements.add("query");
+    statements.add(sql6);
+
+    statements.add("transaction");
+    statements.add("rollback");
+
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sql1), createResultSet(1, "abcd")));
+    mockSpanner.putStatementResult(StatementResult.update(Statement.of(sql2), 1));
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sql3), createResultSet(2, "pqrs")));
+
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sql4), createResultSet(3, "1234")));
+    mockSpanner.putStatementResult(StatementResult.update(Statement.of(sql5), 2));
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sql6), createResultSet(4, "6789")));
+
+    String expectedOutput =
+        "(1, 'abcd')\n" + "1\n" + "(2, 'pqrs')\n" + "(3, '1234')\n" + "2\n" + "(4, '6789')\n";
+    String actualOutput = executeTransactions(pgServer.getLocalPort(), statements);
+    Assert.assertEquals(expectedOutput, actualOutput);
+
+    Assert.assertEquals(1, mockSpanner.countRequestsOfType(CommitRequest.class));
+    Assert.assertEquals(1, mockSpanner.countRequestsOfType(RollbackRequest.class));
+    Assert.assertEquals(6, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+
+    List<ExecuteSqlRequest> requests = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class);
+    List<String> sql = new ArrayList<>();
+
+    for (ExecuteSqlRequest request : requests) {
+      sql.add(request.getSql());
+    }
+
+    Assert.assertEquals(true, sql.contains(sql1));
+    Assert.assertEquals(true, sql.contains(sql2));
+    Assert.assertEquals(true, sql.contains(sql3));
+    Assert.assertEquals(true, sql.contains(sql4));
+    Assert.assertEquals(true, sql.contains(sql5));
+    Assert.assertEquals(true, sql.contains(sql6));
```

Comment:
nit: remove the for loop and replace the `assertEquals(true, ...)` comparisons with an actual comparison of the request at for example index 2 with `sql3`. This then also ensures that the requests were executed in the expected order.

Additional nits here and elsewhere: Prefer `assertTrue(..)` over `assertEquals(true, ...)` and use a static import for `assertEquals` to make the code less verbose.